File: java/lib/src/main/java/io/openweb3/xwebhook/Endpoint.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.exceptions.ApiException;
import io.openweb3.xwebhook.internal.api.EndpointApi;
import io.openweb3.xwebhook.models.EndpointHeadersIn;
import io.openweb3.xwebhook.models.EndpointHeadersOut;
import io.openweb3.xwebhook.models.EndpointHeadersPatchIn;
import io.openweb3.xwebhook.models.EndpointIn;
import io.openweb3.xwebhook.models.EndpointOut;
import io.openweb3.xwebhook.models.EndpointPatch;
import io.openweb3.xwebhook.models.EndpointUpdate;
import io.openweb3.xwebhook.models.EndpointSecretOut;
import io.openweb3.xwebhook.models.EndpointSecretRotateIn;
import io.openweb3.xwebhook.models.EndpointTransformationIn;
import io.openweb3.xwebhook.models.EndpointTransformationOut;
import io.openweb3.xwebhook.models.EventExampleIn;
import io.openweb3.xwebhook.models.ListResponseEndpointOut;
import io.openweb3.xwebhook.models.MessageOut;
import io.openweb3.xwebhook.models.RecoverIn;
import io.openweb3.xwebhook.models.ReplayIn;
import io.openweb3.xwebhook.models.EndpointStats;

public final class Endpoint {
	private final EndpointApi api;

	public Endpoint() {
		api = new EndpointApi();
	}

	public ListResponseEndpointOut list(final String appId, final EndpointListOptions options) throws ApiException {
		try {
			return api.v1EndpointList(appId, options.getLimit(), options.getIterator(), options.getOrder());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointOut create(final String appId, final EndpointIn endpointIn) throws ApiException {
		return this.create(appId, endpointIn, new PostOptions());
	}

	public EndpointOut create(final String appId, final EndpointIn endpointIn, final PostOptions options)
			throws ApiException {
		try {
			return api.v1EndpointCreate(appId, endpointIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointOut get(final String appId, final String endpointId) throws ApiException {
		try {
			return api.v1EndpointGet(appId, endpointId);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointOut update(final String appId, final String endpointId, final EndpointUpdate endpointUpdate)
			throws ApiException {
		try {
			return api.v1EndpointUpdate(appId, endpointId, endpointUpdate);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointOut patch(final String appId, final String endpointId, final EndpointPatch endpointPatch)
			throws ApiException {
		try {
			return api.v1EndpointPatch(appId, endpointId, endpointPatch);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void delete(final String appId, final String endpointId) throws ApiException {
		try {
			api.v1EndpointDelete(appId, endpointId);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointSecretOut getSecret(final String appId, final String endpointId) throws ApiException {
		try {
			return api.v1EndpointGetSecret(appId, endpointId);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void rotateSecret(final String appId, final String endpointId,
			final EndpointSecretRotateIn endpointSecretRotateIn) throws ApiException {
		this.rotateSecret(appId, endpointId, endpointSecretRotateIn, new PostOptions());
	}

	public void rotateSecret(final String appId, final String endpointId,
			final EndpointSecretRotateIn endpointSecretRotateIn, final PostOptions options) throws ApiException {
		try {
			api.v1EndpointRotateSecret(appId, endpointId, endpointSecretRotateIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void recover(final String appId, final String endpointId, final RecoverIn recoverIn) throws ApiException {
		this.recover(appId, endpointId, recoverIn, new PostOptions());
	}

	public void recover(final String appId, final String endpointId, final RecoverIn recoverIn,
			final PostOptions options) throws ApiException {
		try {
			api.v1EndpointRecover(appId, endpointId, recoverIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointHeadersOut getHeaders(final String appId, final String endpointId) throws ApiException {
		try {
			return api.v1EndpointGetHeaders(appId, endpointId);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void updateHeaders(final String appId, final String endpointId, final EndpointHeadersIn endpointHeadersIn)
			throws ApiException {
		try {
			api.v1EndpointUpdateHeaders(appId, endpointId, endpointHeadersIn);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void patchHeaders(final String appId, final String endpointId,
			final EndpointHeadersPatchIn endpointHeadersIn) throws ApiException {
		try {
			api.v1EndpointPatchHeaders(appId, endpointId, endpointHeadersIn);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointStats getStats(final String appId, final String endpointId) throws ApiException {
		return getStats(appId, endpointId, new EndpointStatsOptions());
	}

	public EndpointStats getStats(final String appId, final String endpointId, final EndpointStatsOptions options)
			throws ApiException {
		try {
			return api.v1EndpointGetStats(appId, endpointId, options.getSince(), options.getUntil());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void replayMissing(final String appId, final String endpointId, final ReplayIn replayIn)
			throws ApiException {
		this.replayMissing(appId, endpointId, replayIn, new PostOptions());
	}

	public void replayMissing(final String appId, final String endpointId, final ReplayIn replayIn,
			final PostOptions options) throws ApiException {
		try {
			api.v1EndpointReplay(appId, endpointId, replayIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public EndpointTransformationOut transformationGet(final String appId, final String endpointId)
			throws ApiException {
		try {
			return api.v1EndpointTransformationGet(appId, endpointId);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public void transformationPartialUpdate(final String appId, final String endpointId,
			final EndpointTransformationIn transformationIn) throws ApiException {
		try {
			api.v1EndpointTransformationPartialUpdate(appId, endpointId, transformationIn);
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}

	public MessageOut sendExample(final String appId, final String endpointId, final EventExampleIn eventExampleIn)
			throws ApiException {
		return this.sendExample(appId, endpointId, eventExampleIn, new PostOptions());
	}

	public MessageOut sendExample(final String appId, final String endpointId, final EventExampleIn eventExampleIn,
			final PostOptions options) throws ApiException {
		try {
			return api.v1EndpointSendExample(appId, endpointId, eventExampleIn, options.getIdempotencyKey());
		} catch (io.openweb3.xwebhook.internal.ApiException e) {
			throw Utils.wrapInternalApiException(e);
		}
	}
}

File: kotlin/lib/src/main/kotlin/Endpoint.kt
package io.openweb3.xwebhook.kotlin

import io.openweb3.xwebhook.kotlin.exceptions.ApiException
import io.openweb3.xwebhook.kotlin.internal.apis.EndpointApi
import io.openweb3.xwebhook.kotlin.models.EndpointHeadersIn
import io.openweb3.xwebhook.kotlin.models.EndpointHeadersOut
import io.openweb3.xwebhook.kotlin.models.EndpointHeadersPatchIn
import io.openweb3.xwebhook.kotlin.models.EndpointIn
import io.openweb3.xwebhook.kotlin.models.EndpointOut
import io.openweb3.xwebhook.kotlin.models.EndpointPatch
import io.openweb3.xwebhook.kotlin.models.EndpointSecretOut
import io.openweb3.xwebhook.kotlin.models.EndpointSecretRotateIn
import io.openweb3.xwebhook.kotlin.models.EndpointStats
import io.openweb3.xwebhook.kotlin.models.EndpointTransformationIn
import io.openweb3.xwebhook.kotlin.models.EndpointTransformationOut
import io.openweb3.xwebhook.kotlin.models.EndpointUpdate
import io.openweb3.xwebhook.kotlin.models.EventExampleIn
import io.openweb3.xwebhook.kotlin.models.ListResponseEndpointOut
import io.openweb3.xwebhook.kotlin.models.RecoverIn
import io.openweb3.xwebhook.kotlin.models.ReplayIn

class Endpoint internal constructor(token: String, options: XwebhookOptions) {
    val api = EndpointApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun list(
        appId: String,
        options: EndpointListOptions = EndpointListOptions(),
    ): ListResponseEndpointOut {
        try {
            return api.v1EndpointList(
                appId,
                options.limit,
                options.iterator,
                options.order,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun create(
        appId: String,
        endpointIn: EndpointIn,
        options: PostOptions = PostOptions(),
    ): EndpointOut {
        try {
            return api.v1EndpointCreate(
                appId,
                endpointIn,
                options.idempotencyKey,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun get(
        appId: String,
        endpointId: String,
    ): EndpointOut {
        try {
            return api.v1EndpointGet(endpointId, appId)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun update(
        appId: String,
        endpointId: String,
        endpointUpdate: EndpointUpdate,
    ): EndpointOut {
        try {
            return api.v1EndpointUpdate(
                appId,
                endpointId,
                endpointUpdate,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun patch(
        appId: String,
        endpointId: String,
        endpointPatch: EndpointPatch,
    ): EndpointOut {
        try {
            return api.v1EndpointPatch(
                appId,
                endpointId,
                endpointPatch,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun delete(
        appId: String,
        endpointId: String,
    ) {
        try {
            api.v1EndpointDelete(appId, endpointId)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun getSecret(
        appId: String,
        endpointId: String,
    ): EndpointSecretOut {
        try {
            return api.v1EndpointGetSecret(
                appId,
                endpointId,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun rotateSecret(
        appId: String,
        endpointId: String,
        endpointSecretRotateIn: EndpointSecretRotateIn,
        options: PostOptions = PostOptions(),
    ) {
        try {
            api.v1EndpointRotateSecret(
                appId,
                endpointId,
                endpointSecretRotateIn,
                options.idempotencyKey,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun recover(
        appId: String,
        endpointId: String,
        recoverIn: RecoverIn,
        options: PostOptions = PostOptions(),
    ) {
        try {
            api.v1EndpointRecover(
                appId,
                endpointId,
                recoverIn,
                options.idempotencyKey,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun getHeaders(
        appId: String,
        endpointId: String,
    ): EndpointHeadersOut {
        try {
            return api.v1EndpointGetHeaders(
                appId,
                endpointId,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun updateHeaders(
        appId: String,
        endpointId: String,
        endpointHeadersIn: EndpointHeadersIn,
    ) {
        try {
            api.v1EndpointUpdateHeaders(
                appId,
                endpointId,
                endpointHeadersIn,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun patchHeaders(
        appId: String,
        endpointId: String,
        endpointHeadersIn: EndpointHeadersPatchIn,
    ) {
        try {
            api.v1EndpointPatchHeaders(
                appId,
                endpointId,
                endpointHeadersIn,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun getStats(
        appId: String,
        endpointId: String,
        options: EndpointStatsOptions = EndpointStatsOptions(),
    ): EndpointStats {
        try {
            return api.v1EndpointGetStats(
                appId,
                endpointId,
                options.since,
                options.until,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun replayMissing(
        appId: String,
        endpointId: String,
        replayIn: ReplayIn,
        options: PostOptions = PostOptions(),
    ) {
        try {
            api.v1EndpointReplay(
                appId,
                endpointId,
                replayIn,
                options.idempotencyKey,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun transformationGet(
        appId: String,
        endpointId: String,
    ): EndpointTransformationOut {
        try {
            return api.v1EndpointTransformationGet(
                appId,
                endpointId,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun transformationPartialUpdate(
        appId: String,
        endpointId: String,
        endpointTransformationIn: EndpointTransformationIn,
    ) {
        try {
            api.v1EndpointTransformationPartialUpdate(
                appId,
                endpointId,
                endpointTransformationIn,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun sendExample(
        appId: String,
        endpointId: String,
        eventExampleIn: EventExampleIn,
        options: PostOptions = PostOptions(),
    ) {
        try {
            api.v1EndpointSendExample(
                appId,
                endpointId,
                eventExampleIn,
                options.idempotencyKey,
            )
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
